sdk/src/main/java/com/ifountain/opsgenie/client/AbstractOpsGenieHttpClient.java

```diff
@@ -126,6 +126,57 @@ protected BaseResponse doGetRequest(BaseRequest request) throws OpsGenieClientEx
+        setAndValidateApiKey(request);
+        Map<String, String> headers = request.getReqHeadersForPostOrPatch();
+        String uri = rootUri + request.getEndPoint();
+        String json = JsonUtils.toJson(request);
+        log.info("Executing OpsGenie request to [" + uri + "] with content Parameters:" + LogUtils.getInsensitiveLogMessage(JsonUtils.toMap(request)));
+        OpsGenieHttpResponse httpResponse = httpClient.post(uri, json, headers);
+        handleResponse(httpResponse);
+        return populateResponse(request, httpResponse);
+    }
+    protected BaseResponse doGetRequestV2(BaseRequest request) throws OpsGenieClientException, IOException, ParseException {
+        setAndValidateApiKey(request);
+        String uri = rootUri + request.getEndPoint();
+
+        Map<String,Object> parameters = request.getRequestParams();
+        Map<String,String> headers = request.getReqHeadersForGetOrDelete();
+        log.info("Executing OpsGenie request to [" + uri + "] with Parameters:" + LogUtils.getInsensitiveLogMessage(parameters));
+        OpsGenieHttpResponse httpResponse = httpClient.get(uri, parameters,headers);
+        handleResponse(httpResponse);
+        return populateResponse(request, httpResponse);
+    }
+
+    protected BaseResponse doDeleteRequestV2(BaseRequest request) throws OpsGenieClientException, IOException, ParseException {
+        setAndValidateApiKey(request);
+        String uri = rootUri + request.getEndPoint();
+        Map<String,Object> parameters = request.getRequestParams();
+        Map<String,String> headers = request.getReqHeadersForGetOrDelete();
+        log.info("Executing OpsGenie request to [" + uri + "] with Parameters:" + LogUtils.getInsensitiveLogMessage(parameters));
+        OpsGenieHttpResponse httpResponse = httpClient.delete(uri, parameters,headers);
+        handleResponse(httpResponse);
+        return populateResponse(request, httpResponse);
+    }
+
+    protected BaseResponse doPatchRequestV2(BaseRequest request) throws IOException, OpsGenieClientException, ParseException {
+        setAndValidateApiKey(request);
+        Map<String,Object> parameters = request.getRequestParams();
+        Map<String, String> headers = request.getReqHeadersForPostOrPatch();
+        String uri = rootUri + request.getEndPoint();
+        String json = JsonUtils.toJson(request);
+        log.info("Executing OpsGenie request to [" + uri + "] with content Parameters:" + LogUtils.getInsensitiveLogMessage(JsonUtils.toMap(request)));
+        OpsGenieHttpResponse httpResponse = httpClient.patch(uri, json, parameters,headers);
+        handleResponse(httpResponse);
+        return populateResponse(request, httpResponse);
+    }
+
+    private void setAndValidateApiKey(BaseRequest request) throws OpsGenieClientValidationException {
+        if (request.getApiKey() == null)
+            request.setApiKey(getApiKey());
```

Comment:
from where does this getApiKey() get the api key?

Reply:
api key is a member of this (AbstractOpsGenieHttpClient) class. We have to provide it in the request itself.

Reply:
Example:
        OpsGenieHttpClient opsGenieHttpClient = new OpsGenieHttpClient();
        JsonOpsgenieHttpClient httpClient = new JsonOpsgenieHttpClient(opsGenieHttpClient);
        httpClient.setApiKey("ff898148-247e-4055-8ec0-d5573ea715a1");